MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,148 @@
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Returns a <see cref="IEnumerable{T}"/> that lazily creates an in-memory
+        /// cache of the enumeration on first iteration, if it is not already an
+        /// in-memory source.
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <returns>Returns a sequence that corresponds to a cached version of the input sequence.</returns>
+        public static IEnumerable<T> Memoize<T>(this IEnumerable<T> source)
+        {
+            return source.Memoize(false, false);
+        }
+
+        /// <summary>
+        /// Returns a <see cref="IEnumerable{T}"/> that lazily creates an in-memory
+        /// cache of the enumeration on first iteration, if it is not already an
+        /// in-memory source.
+        /// An additional argument specifies if buffering must happen even if the source implements
+        /// <see cref="ICollection{T}"/> (otherwise it is assumed that the source already serves as 
+        /// a buffer adequately).
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="forceBuffering">Force buffering, even if source is an <see cref="ICollection{T}" />
+        /// (otherwise source is assumed to already be in-memory and enumerated quickly, and hence that
+        /// buffering again would be a waste).</param>
+        /// <param name="disposeOnEarlyExit">Indicates if the call to dispose method of source's enumerator, 
+        /// and therefore the close of the buffering, must happen at the end of the first iteration (true) 
+        /// or only when source is entirely iterated (false).</param>
+        /// <returns>Returns a sequence that corresponds to a cached version of the input sequence.</returns>
+        public static IEnumerable<T> Memoize<T>(this IEnumerable<T> source, bool forceBuffering, bool disposeOnEarlyExit)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            if (!forceBuffering && source is ICollection<T>)
+            {
+                return source;
+            }
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source, disposeOnEarlyExit);
```

Comment:
There is a test missing for this optimisation. If I simple rewrite the line as:

```c#
return new MemoizedEnumerable<T>(source, disposeOnEarlyExit);
```

then all tests still pass.